File: fates/enums.py
from aenum import IntEnum, StrEnum as Enum


class WidgetFormat(Enum):
    """Widget format"""

    _init_ = "value __doc__"

    json = "json", "Raw JSON format"

    html = "html", "HTML format"

    png = "png", "PNG image format. May produce better results than WEBP"

    webp = "webp", "WebP image format"


class BotServerFlag(IntEnum):
    """Flags that apply to both bots and servers"""

    _init_ = "value __doc__"

    Unlocked = 0, "Bot or server is unlocked and can be freely editted"

    EditLocked = (
        1,
        "Bot or server is locked and cannot be editted but can be unlocked by the owner",
    )

    StaffLocked = (
        2,
        "Bot or server is locked by staff and can only be unlocked by staff",
    )

    StatsLocked = 3, "Bot or server stats are locked due to abuse"

    VoteLocked = 4, "Bot or server is locked from voting due to abuse"

    System = 5, "Bot or server is a system bot or server"

    WhitelistOnly = (
        6,
        "Server is a whitelist only server (it cannot be joined without being whitelisted by staff on said server)",
    )

    KeepBannerDecor = (
        7,
        "Bot or server banner should keep special fallback banner styles",
    )

    NSFW = 8, "Bot or server is NSFW"

    LoginRequired = 9, "Server requires the user to be logged in to join"


class UserFlag(IntEnum):
    """Flags that apply to users"""

    _init_ = "value __doc__"

    Unknown = 0, "Unknown flag"

    VotesPrivate = 1, "User's votes are private"

    Staff = 2, "User is a staff member"

    AvidVoter = 3, "User is an avid voter"


class UserState(IntEnum):
    """User states"""

    _init_ = "value __doc__"

    Normal = 0, "User is normal (not banned)"

    GlobalBan = 1, "User is globally banned from the list (apart from data actions)"

    ProfileEditBan = 2, "User is banned from editing their profile due to abuse"


class BotServerState(IntEnum):
    """Bot or server state"""

    Approved = 0
    Pending = 1
    Denied = 2
    Hidden = 3
    Banned = 4
    UnderReview = 5
    Certified = 6
    Archived = 7
    PrivateViewable = 8
    PrivateStaffOnly = 9


class UserBotAction(IntEnum):
    """Actions a user can take on a bot"""

    Approve = 0
    Deny = 1
    Certify = 2
    Ban = 3
    Claim = 4
    Unclaim = 5
    TransferOwnership = 6
    EditBot = 7
    DeleteBot = 8
    Unban = 9
    Uncertify = 10
    Unverify = 11
    Requeue = 12


class LongDescriptionType(IntEnum):
    """The type of long description. This is used to determine how the long description should be processed/rendered"""

    _init_ = "value __doc__"

    Html = 0, "Raw HTML. No markdown processing is applied"

    MarkdownServerSide = (
        1,
        "Markdown. Markdown is processed server side using the ``cmarkgfm`` library.",
    )


class WebhookType(IntEnum):
    """The type of webhook (Fates Client is deprecated and has no effect anymore)"""

    Vote = 0
    DiscordIntegration = 1
    DeprecatedFatesClient = 2  # There in case any bots are still using it


class CommandType(IntEnum):
    """The type of bot command"""

    PrefixCommand = 0
    SlashCommandGlobal = 1
    SlashCommandGuild = 2


class TargetType(IntEnum):
    """A bot/server/user etc."""

    Bot = 0
    Server = 1
    User = 2
    Pack = 3


class PageStyle(IntEnum):
    """Page style for the bot/server page"""

    _init_ = "value __doc__"

    Tabs = 0, "Legacy tabs style"

    SingleScroll = 1, "Single scroll/no tabs style"


class UserExperiment(IntEnum):
    """User experiments"""

    _init_ = "value __doc__"

    Unknown = 0, "Unknown experiment"

    GetRoleSelector = 1, "We switched to native roles. This experiment FAILED"

    LynxExperimentRolloutView = 2, "Lynx experiment rollout view. No longer relevant"

    BotReport = 3, "Bot reports are supported"

    ServerAppealCertification = (
        4,
        "Ability to use request type of Appeal or Certification in server appeal",
    )

    UserVotePrivacy = (
        5,
        "The ability for users to hide their votes from Get Bot Votes and Get Server Votes API",
    )

    DevPortal = 6, "The ability to use the developer portal. No longer relevant"


class DataAction(Enum):
    """Data actions"""

    Request = "req"

    Delete = "del"

File: sunbeam/src/lib/enums/enums.ts
enum BotState {
  approved = 0,
  pending = 1,
  denied = 2,
  hidden = 3,
  banned = 4,
  under_review = 5,
  certified = 6,
  archived = 7,
  private_viewable = 8,
  private_staff_only = 9
}

enum UserState {
  normal = 0,
  global_ban = 1,
  profile_edit_ban = 2,
  ddr_ban = 3, // Unused
  api_ban = 4 // Unused
}

enum UserFlags {
  Unknown = 0,
  VotesPrivate = 1,
  Staff = 2,
  AvidVoter = 3,
  Premium = 4
}

// This only covers site-supported user experiments
enum UserExperiments {
  Unknown = 0,
  GetRolesSelector = 1,
  LynxExperimentRolloutView = 2,
  BotReport = 3,
  ServerAppealCertification = 4,
  VotePrivacy = 5,
  DevPortal = 6
}

enum UserBotAction {
  approve = 0,
  deny = 1,
  certify = 2,
  ban = 3,
  claim = 4,
  unclaim = 5,
  transfer_ownership = 6,
  edit_bot = 7,
  delete_bot = 8,
  unban = 9,
  uncertify = 10,
  unverify = 11,
  requeue = 12
}

enum ReviewType {
  bot = 0,
  server = 1
}

enum CommandType {
  regular = 0,
  guild_slash = 1,
  global_slash = 2
}

enum Flag {
  unlocked = 0,
  edit_locked = 1,
  staff_locked = 2,
  stats_locked = 3,
  vote_locked = 4,
  system = 5,
  whitelist_only = 6,
  keep_banner_decor = 7,
  nsfw = 8
}

enum PageStyle {
  tabs = 0,
  single_view = 1
}

// Alert Types
export enum AlertType {
  Error = 1,
  Warning = 2,
  Info = 3,
  Alert = 4,
  Success = 5,
  Debug = 6,
  Prompt = 7
}

export enum AlertInputType {
  Text = 1,
  Number = 2,
  Boolean = 3,
  DateTime = 4,
  DateTimeLocal = 5,
  Color = 6,
  File = 7,
  Pre = 8 // Implements a metadata html element that can be put before another input
}

export enum TargetType {
  Bot = 0,
  Server = 1,
  User = 2,
  Pack = 3
}

export enum SettingsMode {
  Add = 1,
  Edit = 2
}

export enum UserStatus {
  UserStatusOnline = 0,
  UserStatusIdle = 1,
  UserStatusDnd = 2,
  UserStatusOffline = 3
}

export const enums = {
  BotState: BotState,
  SettingsMode: SettingsMode,
  UserState: UserState,
  UserBotAction: UserBotAction,
  CommandType: CommandType,
  ReviewType: ReviewType,
  PageStyle: PageStyle,
  AlertType: AlertType,
  AlertInputType: AlertInputType,
  Flags: Flag,
  UserFlags: UserFlags,
  UserExperiments: UserExperiments,
  UserStatus: UserStatus,
  LongDescType: {
    html: 0,
    markdown_server: 1
  },
  TargetType: TargetType,
  helpers: {
    flagCheck: function (flags: number[] | number, allFlags: number[]) {
      // Function start
      let flagsParsed: number[] = [];
      if (typeof flags === 'number') {
        flagsParsed.push(flags);
      } else {
        flagsParsed = flags;
      }

      return flagsParsed.some((item) => allFlags.includes(item));
      // End of function
    },
    targetTypeString: function (type: TargetType) {
      switch (type) {
        case TargetType.Bot:
          return 'bot';
        case TargetType.Server:
          return 'server';
        case TargetType.User:
          return 'user';
        case TargetType.Pack:
          return 'pack';
      }
    },
    strToTargetType: function (type: string) {
      switch (type) {
        case 'bot':
          return TargetType.Bot;
        case 'server':
          return TargetType.Server;
        case 'user':
          return TargetType.User;
        case 'pack':
          return TargetType.Pack;
      }
    }
  }
};

// Interfaces

export interface DiscordUser {
  id: string;
  username: string;
  disc: string;
  avatar: string;
  bot: boolean;
  system: boolean;
  status: UserStatus;
  flags: number;
}

export interface Snippet {
  user: DiscordUser;
  votes: number;
  description: string;
  flags: Flag[];
  banner_card: string;
  state: BotState;
  guild_count: number;
}

export interface Index {
  new: Snippet[];
  top_voted: Snippet[];
  certified: Snippet[];
}

export interface Tag {
  id: string;
  name: string;
  iconify_data: string;
  owner_guild: string;
}

export interface Feature {
  id: string;
  name: string;
  viewed_as: string;
  description: string;
}

export interface BotMeta {
  tags: Tag[];
  features: Feature[];
}

export interface ServerMeta {
  tags: Tag[];
}

export interface ListMeta {
  bot: BotMeta;
  server: ServerMeta;
}
